File: client/src/features/explore/services/ExploreService.js
import Api from "../../../api/Api";

class ExploreService {
    constructor() {
        this.base = "/explore";
    }

    // ── Single call: options + all country data + saved prefs ──────────────
    async getOptions() {
        await new Promise(resolve => setTimeout(resolve, 300));
        const response = await Api.get(`${this.base}/options`);
        return response; // ✅ Already returns the full response
    }

    // ── Get Explore feed ───────────────────────────────────────────────────
    // In ExploreService.js

    async getExplore(filters = {}) {
        console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
        console.log('📤 FETCHING EXPLORE PROFILES');
        console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
        console.log('Filters being sent to backend:', JSON.stringify(filters, null, 2));

        const params = new URLSearchParams();

        // Basic filters
        if (filters.gender) params.append("gender", filters.gender);
        if (filters.city) params.append("city", filters.city);

        // ✅ Handle single country OR multiple countries
        if (filters.country) {
            params.append("country", filters.country);
        } else if (filters.countries) {
            // Multiple countries as comma-separated string
            params.append("countries", filters.countries);
        }

        if (filters.minAge) params.append("minAge", filters.minAge);
        if (filters.maxAge) params.append("maxAge", filters.maxAge);

        // Additional filters from activeFilters
        if (filters.isVerified) params.append("isVerified", "1");
        if (filters.isPremium) params.append("isPremium", "1");
        if (filters.isOnline) params.append("isOnline", "1");

        const query = params.toString();
        console.log('Query string:', query || '(no filters)');
        console.log('Full URL:', `${this.base}/get-explore${query ? `?${query}` : ""}`);
        console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

        return Api.get(`${this.base}/get-explore${query ? `?${query}` : ""}`);
    }
    // ── Send Interest (Like) ───────────────────────────────────────────────
    async sendInterest(toUserId, isSuperLike = false) {
        return Api.post(`/interest/send-interest`, {
            interestId: Number(toUserId),
            isSuperLike,
        });
    }

    // ── Send Dislike (Pass) ────────────────────────────────────────────────
    async sendDislike(targetUserId) {
        return Api.post(`/interest/dislike`, { interestId: Number(targetUserId) });
    }

    // ── Save preferences ───────────────────────────────────────────────────
    async savePreferences(payload) {
        try {
            const response = await Api.post(`${this.base}/save-preferences`, payload);



            // ✅ Return response.data if it exists, otherwise the whole response
            return response?.data || response;

        } catch (error) {
            console.error('❌ savePreferences error:', error);

            // ✅ Extract meaningful error message
            const errorMessage = error.response?.data?.message
                || error.response?.data?.error
                || error.message
                || 'Failed to save preferences';

            // ✅ Throw with the message so the caller can catch it
            throw new Error(errorMessage);
        }
    }
}

export default new ExploreService();

File: app/lib/features/explore/services/explore_service.dart
import 'package:get/get.dart';
import '../../../data/providers/api_client.dart';

class ExploreService extends GetxService {
  final ApiClient _apiClient = Get.find<ApiClient>();

  // Get explore feed with filters
  Future<Map<String, dynamic>?> getExplore({
    String? gender,
    String? city,
    String? country,
    String? countries,
    int? minAge,
    int? maxAge,
    bool isVerified = false,
    bool isPremium = false,
    bool isOnline = false,
  }) async {
    final params = <String, String>{};
    if (gender != null && gender.isNotEmpty) params['gender'] = gender;
    if (city != null && city.isNotEmpty) params['city'] = city;
    if (country != null && country.isNotEmpty) params['country'] = country;
    if (countries != null && countries.isNotEmpty) params['countries'] = countries;
    if (minAge != null) params['minAge'] = minAge.toString();
    if (maxAge != null) params['maxAge'] = maxAge.toString();
    if (isVerified) params['isVerified'] = '1';
    if (isPremium) params['isPremium'] = '1';
    if (isOnline) params['isOnline'] = '1';

    return await _apiClient.get('/explore/get-explore', queryParameters: params);
  }

  // Get options (countries, religions, etc)
  Future<Map<String, dynamic>?> getOptions() async {
    return await _apiClient.get('/explore/options');
  }

  // Send interest (like)
  Future<Map<String, dynamic>?> sendInterest(int toUserId, {bool isSuperLike = false}) async {
    return await _apiClient.post('/interest/send-interest', data: {
      'interestId': toUserId,
      'isSuperLike': isSuperLike,
    });
  }

  // Send dislike (pass)
  Future<Map<String, dynamic>?> sendDislike(int targetUserId) async {
    return await _apiClient.post('/interest/dislike', data: {
      'interestId': targetUserId,
    });
  }

  // Save preferences
  Future<Map<String, dynamic>?> savePreferences(Map<String, dynamic> payload) async {
    return await _apiClient.post('/explore/save-preferences', data: payload);
  }
}
